clash-lib/src/Clash/Normalize/Transformations.hs

```diff
@@ -424,9 +424,63 @@ caseCase _ e = return e
 {-# SCC caseCase #-}
 
 -- | Inline function with a non-representable result if it's the subject
--- of a Case-decomposition
+-- of a Case-decomposition. It's a custom topdown traversal that -for efficiency
+-- reasons- does not explore alternative of cases whose subject triggered an
+-- 'inlineNonRepWorker'.
 inlineNonRep :: HasCallStack => NormRewrite
-inlineNonRep _ e@(Case scrut altsTy alts)
+inlineNonRep _ctx = fmap snd . go
+ where
+  go :: HasCallStack => Term -> NormalizeSession (Bool, Term)
+  go e0 = inlineNonRepWorker e0 >>= \case
+    (True, e1) ->
+      return (True, e1)
+    (False, e1) ->
+      case e1 of
+        Var _ -> return (False, e1)
+        Data _ -> return (False, e1)
+        Literal _ -> return (False, e1)
+        Prim _ -> return (False, e1)
+        Lam v t -> second (Lam v) <$> go t
+        TyLam v t -> second (TyLam v) <$> go t
+        App a0 b0 -> do
+          (change0, a1) <- go a0
+          (change1, b1) <- go b0
+          return (change0 || change1, App a1 b1)
+        TyApp t typ -> second (`TyApp` typ) <$> go t
+        Cast t0 typ0 typ1 -> do
+          (change, t1) <- go t0
+          return (change, Cast t1 typ0 typ1)
+        Tick tickInfo t -> second (Tick tickInfo) <$> go t
+        Letrec bindings_ body0 -> do
+          let (bndrs, bodies0) = unzip bindings_
+          (changes, bodies1) <- unzip <$> mapM go bodies0
+          (change, body1) <- go body0
```

Comment:
Now I'm definitely more in favour of ditching `Bool` and using `Writer` to find changes. Think how much nicer it would make these cases to not have to collect / propagate whether things have changed